lineapy/graph_reader/session_artifacts.py

```diff
@@ -413,7 +433,71 @@ def _slice_session_artifacts(self) -> None:
         )
         self.input_parameters_nodecollection._update_graph(self.graph)
 
+    def _update_nodecollection_dependencies(self):
+        """
+        Update nodecollection dependencies when we replace the cached artifact
```

Comment:
I'd argue this function name and comment are misleading because `nodecollection_dependencies` is not actually used anywhere else in this function and its update is really only used to facilitate changes to `artifact_nodecollections`. 

If my understanding is correct, the actual important piece of this function is that it updates `artifact_nodecollections` which is used extensively in code generation by SessionArtifact. 

Lets clarify in the comments here that this function uses `nodecollection_dependencies` to help update 
`artifact_nodecollections`, and that the reason for this update is so that we can cut out any code blocks from node collections that are not useful in generating output artifacts. These cases arise when some node collections are not connected to output_artifacts when input_artifacts are removed from the graph, this means that these node collections are not needed and can be dropped. 

Its quite wordy, but I don't think this concept is trivial, and this is the function that does it so we should spend some time clarifying how and why this is done.